QGIS 3 PyQGIS Cookbook update (#3406)
qgis/QGIS-Documentation: source/docs/pyqgis_developer_cookbook/canvas.rst

Whenever the map has been panned, zoomed in/out (or some other action that triggers
a refresh), the map is rendered again within the current extent. The layers are

> rendered to an image (using :class:`QgsMapRendererJob <qgis.gui.QgsMapRendererJob>` class) and that image is

Reviewer: ```suggestion
rendered to an image (using the :class:`QgsMapRendererJob <qgis.gui.QgsMapRendererJob>` class) and that image is
```

> rendered to an image (using :class:`QgsMapRendererJob <qgis.gui.QgsMapRendererJob>`_ class) and that image is

Reviewer: ```suggestion
rendered to an image (using :class:`QgsMapRendererJob <qgis.gui.QgsMapRendererJob>` class) and that image is
```

Reviewer: Is it worth to test this?

>  displayed on the canvas. The :class:`QgsMapCanvas <qgis.gui.QgsMapCanvas>`  class also controls refreshing

Reviewer: ```suggestion
displayed on the canvas. The :class:`QgsMapCanvas <qgis.gui.QgsMapCanvas>`  class also controls refreshing
```

> then displayed on the canvas. The :class:`QgsMapCanvas <qgis.gui.QgsMapCanvas>`_  class also controls refreshing

Reviewer: ```suggestion
then displayed on the canvas. The :class:`QgsMapCanvas <qgis.gui.QgsMapCanvas>`  class also controls refreshing
```

Reviewer: Is it worth to test this?

of the rendered map. Besides this item which acts as a background, there may be
more **map canvas items**.
Typical map canvas items are rubber bands (used for measuring, vector editing
etc.) or vertex markers. The canvas items are usually used to give visual
feedback for map tools, for example, when creating a new polygon, the map tool
creates a rubber band canvas item that shows the current shape of the polygon.

> All map canvas items are subclasses of :class:`QgsMapCanvasItem <qgis.gui.QgsMapCanvasItem>`_  which adds

Reviewer: ```suggestion
All map canvas items are subclasses of :class:`QgsMapCanvasItem <qgis.gui.QgsMapCanvasItem>` which adds
```

The following example constructs a window that contains a map canvas and basic
map tools for map panning and zooming. Actions are created for activation of

> each tool: panning is done with :class:`QgsMapToolPan <qgis.gui.QgsMapToolPan>`_, zooming in/out with a

Reviewer: ```suggestion
each tool: panning is done with :class:`QgsMapToolPan <qgis.gui.QgsMapToolPan>`, zooming in/out with a
```

> pair of :class:`QgsMapToolZoom <qgis.gui.QgsMapToolZoom>`_ instances. The actions are set as checkable and

Reviewer: ```suggestion
pair of :class:`QgsMapToolZoom <qgis.gui.QgsMapToolZoom>` instances. The actions are set as checkable and
```

later assigned to the tools to allow automatic handling of checked/unchecked
state of the actions -- when a map tool gets activated, its action is marked as
selected and the action of the previous map tool is deselected. The map tools

> are activated using :func:`setMapTool() <qgis.gui.QgsMapCanvas.setMapTool>`_ method.

Reviewer: ```suggestion
are activated using :func:`setMapTool() <qgis.gui.QgsMapCanvas.setMapTool>` method.
```

.. include:: snippets/canvas.rst_4.py
  :code: python

You can try the above code in the Python console editor. To invoke the canvas window, 

> add also the following lines to instantiate the ``MyWnd``class. They will put the currently 

Reviewer: ```suggestion
add the following lines to instantiate the ``MyWnd``class. They will render the currently 
```

> selected layer into the newly created canvas

Reviewer: ```suggestion
selected layer on the newly created canvas
```

To show some additional data on top of the map in canvas, use map canvas items.
It is possible to create custom canvas item classes (covered below), however
there are two useful canvas item classes for convenience:

> :class:`QgsRubberBand <qgis.gui.QgsRubberBand>`_ for drawing polylines or polygons, and

Reviewer: ```suggestion
:class:`QgsRubberBand <qgis.gui.QgsRubberBand>` for drawing polylines or polygons, and
```

> :class:`QgsVertexMarker <qgis.gui.QgsVertexMarker>`_ for drawing points. They both work with map

Reviewer: ```suggestion
:class:`QgsVertexMarker <qgis.gui.QgsVertexMarker>` for drawing points. They both work with map
```

Rubber bands allow some customization, namely to change their color and line
width

.. include:: snippets/canvas.rst_8.py
  :code: python

The canvas items are bound to the canvas scene. To temporarily hide them (and
show them again), use the :func:`hide` and :func:`show` combo. To completely remove
the item, you have to remove it from the scene of the canvas

.. include:: snippets/canvas.rst_9.py
  :code: python

(in C++ it's possible to just delete the item, however in Python ``del r``
would just delete the reference and the object will still exist as it is owned
by the canvas)

Rubber band can be also used for drawing points, however

> :class:`QgsVertexMarker <qgis.gui.QgsVertexMarker>`_ class is better suited for this

Reviewer: ```suggestion
:class:`QgsVertexMarker <qgis.gui.QgsVertexMarker>` class is better suited for this
```

> (:class:`QgsRubberBand <qgis.gui.QgsRubberBand>`_` would only draw a rectangle around the desired point).

Reviewer: ```suggestion
(:class:`QgsRubberBand <qgis.gui.QgsRubberBand>` would only draw a rectangle around the desired point).
```

Reviewer: This seems to a typo (extra `)

> Here's how to use the vertex marker:

Reviewer: ```suggestion
You can use the vertex marker like this:
```

.. include:: snippets/canvas.rst_10.py
  :code: python

This will draw a red cross on position [0,0]. It is possible to customize the

> icon type, size, color and pen width

Reviewer: ```suggestion
icon type, size, color and pen width.
```

You can write your custom tools, to implement a custom behavior to actions
performed by users on the canvas.

> Map tools should inherit from the :class:`QgsMapTool <qgis.gui.QgsMapTool>`_, class or any derived

Reviewer: ```suggestion
Map tools should inherit from the :class:`QgsMapTool <qgis.gui.QgsMapTool>`, class or any derived
```

> class, and selected as active tools in the canvas using the :func:`setMapTool() <qgis.gui.QgsMapCanvas.setMapTool>`_

Reviewer: ```suggestion
class, and selected as active tools in the canvas using the :func:`setMapTool() <qgis.gui.QgsMapCanvas.setMapTool>`
```
